### PR TITLE
Migration to the new .csproj format

### PR DESCRIPTION
**Pull request by Federerer**
Work in progress

### CODE REVIEW

File: Circuit/Circuit.csproj.old [hunk 1]
```diff
@@ -0,0 +1,134 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dsharlet) [inline Circuit/Circuit.csproj.old:1]: Remove this .old file?

**Review by dsharlet — COMMENTED** [review]
Thanks for this, this looks mostly good, just two recurring comments:
- Can you remove the ".csproj.old" files? I assume they aren't used for anything?
- There are some seemingly irrelevant changes, I think it would be good to revert if they aren't needed, mainly in the unsafe buffer conversion code.

**Review by dsharlet — COMMENTED** [review]
Another thing I noticed looking at the ComputerAlgebra changes, it looks like all the AssemblyInfo and other properties were removed in that PR. Should that happen here too?

**Comment by Federerer** [comment]
The csproj.old files will be removed. I will remove them as I'll be done with the project they belong to.  AsemmblyInfos could potentially stay (with `GenerateAssemblyInfo set to false in .csproj) but i'll remove them, because they can now be generated automatically.

File: Circuit/Circuit.csproj.old (hunk 1 above)
Comment (Federerer) [reply Circuit/Circuit.csproj.old]: I've made this changes, to avoid using P/Invoke Windows-specific implementations in Audio library. Parameter naming might be reverted if you like. I've changed that for API consistency, and to follow C# naming conventions (using the scout rule) ;)

**Comment by Federerer** [comment]
I've made some changes. AssemblyInfos still left to be done.
I'll probably update .net framework to the latest version (4.8) too.

**Comment by dsharlet** [comment]
If the AssemblyInfo.cs are still used, it's fine to leave that for a separate change or I can do it. I was just wondering if they were actually unused after changing the project file format.

In general, I prefer to make more smaller changes that are independent when possible, so let's try to keep the change focused on the project files and any changes necessary because of that :)

**Comment by Federerer** [comment]
Ok, it should be good now. You can now check if everything still works as before ;) Using fresh copy of the repository is recommended, because Visual Studio complains when you switch between branches with different project structure and versions (sometimes it requires to manually remove obj and bin folders to compile successfully).

**Comment by dsharlet** [comment]
I'm testing this and found some interesting things so far:

- Some things seem *dramatically* faster with net50/netcoreapp31 vs. net48. The MXR Phase 90 example (slowest to solve) solves in 1.5s instead of 3s and runs a bit faster too. The Bassman preamp example runs a *lot* faster (1.7x realtime vs. 1.1x). Many circuits seem to run 50%-2x faster!
- It changes the output structure from bin\Release\* to bin\Release\{.net framework}\*, which messes with some workflow things, and is kind of confusing. I can't tell when I "run" from VS, which configuration it is actually using?
- Which one *should* we use? The factors that I am considering are:
-- I want SIMD support. Seems to be in net50/netcoreapp3.1, but net50 seems to be in beta?
-- Does .net core support WPF? Seems like it? .net core 3.1 seems like it is not in beta, and I think it supports SIMD?
-- bin\Release\netcoreapp3.1\LiveSPICE.exe seems to work fine... so it seems like WPF does work?

It seems like we should just go with .net core 3.1 always? I am struggling to find a downside to it...

**Comment by dsharlet** [comment]
Another thing: assuming the above is correct, I'd like to just always target .net core 3.1, and not have it be a subfolder of 'Release'. What do you think about that? I'd rather not have parallel sets of binaries to get confused by (e.g. I don't know which one VS is running with debug mode), and I'd rather not have to change the installer and other scripts I have as part of my workflow (and presumably change them again when .net core 3.2/4.0 come around).

**Comment by Federerer** [comment]
For debugging, you can switch between versions here:
image
We should probably stick to one version in release builds, I've added multi-targeting just for development, to easily compare between versions. There will not be .net core 3.2/4.0, .net and .net core are now merged together to  .net 5.0 which is basically new version of .net core ;) WPF works in all three frameworks I've added. I sometimes use .net 4.8 to extract compiled lambda using dotPeek and CompileToMethod API, which is not available in .net core yet.

**Comment by dsharlet** [comment]
I see, so what do you think the right thing to do is? It seems like maybe .net 4.8 and .net 5.0 are the relevant ones going forward? Also I see .net 5.0 was just announced a few days ago so I guess it's not beta any more.

It *was* useful to compare .net 4.8 vs. .net core 3.1, which really shows amazing performance improvements. Solving circuits and running them is 50%-2x faster in almost all cases!! It might actually mean that using .net 4.8 to inspect the lambdas is misleading, because the compiled lambdas are clearly very different...

Basically the one thing I'd like to resolve before merging is if we need to keep multitargeting. If so, I'll merge it and update all the scripts that need the paths to the build directories.

**Comment by dsharlet** [comment]
Another interesting observation: there is a minor performance difference between .net 5.0 windows and .net core 3.1: it seems that solving circuits is a little faster with .net 5.0 windows, but running them is *slightly* faster with .net core 3.1, so there is an interesting tradeoff there :)

Still, the difference is tiny compared to the difference with .net 4.8 and either of them. I don't think it's worth considering in the decision of which to use.

**Comment by Federerer** [comment]
I think that we should keep 4.8 and 5.0,  having the classic .net framework might be sometimes useful. I've added net core 3.1 before 5.0 was released. For the build scripts, updating them to use the .net 5.0 build only should be fine, as there are no difference in supported platforms between 4.8 and 5.0 (Windows 7 and up)

**Comment by dsharlet** [comment]
OK, so this is ready to merge then it seems?

 Removing .net core 3.1 doesn't seem like it matters that much. As long as we want multitargeting, I'm fine with leaving it.

**Comment by dsharlet** [comment]
Actually, it looks like the VST plugin projects are still only targeting .net 4.8.

**Comment by dsharlet** [comment]
I ran into another issue: I can run LiveSPICE.exe from the build directory and VS, but when I run it after building + running the installer, it tells me I need to install .NET. Even after downloading and installing the .net 5.0 runtime (in addition to the developer package), it still says this.

**Comment by dsharlet** [comment]
OK, the problem with that was some .json files in the build directory need to be shipped in the installer.

**Comment by Federerer** [comment]
I forgot about that, there is a problem with VST plugin with .net core builds. Some of the UI related assemblies are not loaded correctly by the bridge dll, but I didn't had enough time to investigate it further. Could we leave it as is here, and create another issue to handle this? I did't tested installer related stuff as it looks like the build scripts are not available in the repository ;)

**Comment by dsharlet** [comment]
The installer script is at the root of the repo: LiveSPICESetup.iss. I have a few local modifications to it while trying to get it to work:
- Changed paths from Release\* to Release\net50-windows\*
- Added *.json in addition to *.dll

This has enabled LiveSPICE to work, but as you point out, I see this from the VST plugin:
```
Could not load file or assembly 'PresentationFramework, Version=5.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at SharpSoundDevice.PluginLoader.CreatePluginInstance(Assembly asm)
   at SharpSoundDevice.PluginLoader.Create(String assemblyFilename, String bridgeDllDir)
   at SharpSoundDevice.Interop.CreateDevice(String dllFilename, String assemblyFilename)
   at CreateDevice()
   at AudioDevice.{ctor}(AudioDevice* )
   at VstPluginBridge.{ctor}(VstPluginBridge* , IntPtr audioMaster)
```
It does work in MockVst at least... @mikeoliphant any ideas on this issue?

I'm a bit nervous leaving the VST plugin on .net 4.8, because it means there are a lot of differences in behavior between LiveSPICE and the VST plugin. Performance is one, but that's just an indicator that the code running is very different. E.g. maybe there are different numerical stability behaviors between the two?

**Comment by mikeoliphant** [comment]
The SharpSoundDevice bridge dll is built using .NET Framework.

**Comment by dsharlet** [comment]
So, it's kind of disgusting, but it seems to work if we simply ship the .net48 version of the GUI libraries, but .net50 versions of Circuit.dll, ComputerAlgebra.dll. Hopefully that gives us the same circuit simulation code as .net50?

**Comment by dsharlet** [comment]
Hmm, it seems like maybe it's working, but not actually using the .net 5.0 framework. I think the performance is the same as before when running the VST plugin.

**Comment by dsharlet** [comment]
OK I give up on this for now. I think we should merge this, because the issue is orthogonal. We can keep using .net 4.8 for both LiveSPICE and the VST plugin while upgrading to the new project format, and we can separately work on getting the VST plugin to use .net 5.0.

**Comment by dsharlet** [comment]
I just took another final skim through and noticed two things:

- The VST plugin still has .csproj.old files
- A lot of the core assemblies (Circuit, etc.) have "TargetFramework" set to .net20standard, which is different from the rest... is this right?

**Comment by dsharlet** [comment]
I've been doing some reading on the last point. It seems like which .net version we target for the libraries shouldn't affect how the code runs, only which APIs are available. Now the thing I'm confused about is why targeting .net 5 is so much faster when the target of the assembly doing the processing (Circuit.dll) is the same.

**Comment by dsharlet** [comment]
As far as I can tell, the top level executable determines the runtime, and then the rest of the libraries that are loaded just need to be compatible with that? So it doesn't matter which version of the libraries we grab for the installer, only the top level executable. In the case of the VST plugin, I guess that's SharpSoundDevice.

**Comment by dsharlet** [comment]
Just going to merge this, I think I've sorted out the .net version issues enough. Thanks for doing this, I know it must have been a pain. I'll delete the .csproj.old files.

**Comment by Federerer** [comment]
> Hmm, it seems like maybe it's working, but not actually using the .net 5.0 framework. I think the performance is the same as before when running the VST plugin.

> It seems like which .net version we target for the libraries shouldn't affect how the code runs, only which APIs are available. Now the thing I'm confused about is why targeting .net 5 is so much faster when the target of the assembly doing the processing (Circuit.dll) is the same.

> So it doesn't matter which version of the libraries we grab for the installer, only the top level executable. In the case of the VST plugin, I guess that's SharpSoundDevice.

Yes, you basically got it right. .NET standard just defines a set of API's used, so that guarantees that they can be used cross-platform using any runtime supporting that standard. That's why I removed some of the P/Invokes replacing them with Unsafe nuget package. In runtime windows CLR should replace them back with native function calls (but different, depending on platform). This way C# code can be sometimes more efficient than native code compiled to work on different hardware platforms, because JIT compiler can compile using all available hardware features of the currently used machine.
